File: 2021/10/30/RemoveDuplicates.java
/**
 * https://leetcode.com/problems/remove-all-adjacent-duplicates-in-string/
 * @author Andrew Jarombek
 * @since 10/30/2021
 */

import java.util.ArrayDeque;
import java.util.Deque;

public class RemoveDuplicates {
    public static String removeDuplicates(String s) {
        StringBuilder sb = new StringBuilder();
        Deque<Character> deque = new ArrayDeque<>();

        for (char c : s.toCharArray()) {
            deque.addLast(c);
        }

        while (deque.size() > 0) {
            Character c1 = deque.removeFirst();
            Character c2 = null;

            if (deque.size() > 0) {
                c2 = deque.removeFirst();
            }

            if (c2 == null || c1 != c2) {
                if (sb.length() > 0 && sb.charAt(sb.length() - 1) == c1) {
                    sb.deleteCharAt(sb.length() - 1);
                } else {
                    sb.append(c1);
                }

                if (c2 != null)
                    deque.addFirst(c2);
            }
        }

        return sb.toString();
    }

    public static void main(String[] args) {
        String result = removeDuplicates("aababaab");
        assert result.equals("ba");
    }
}

File: 2022/07/05/removeDuplicates.ts
/**
 * https://leetcode.com/problems/remove-all-adjacent-duplicates-in-string/
 * @author Andrew Jarombek
 * @since 7/5/2022
 */

// Attempt #1 - Array (timeout for large strings)
function removeDuplicates(s: string): string {
    let chars = s.split("");
    let end = s.length - 1;

    while (end > 0) {
        let i = 0;

        while (i < end) {
            if (chars[i] == chars[i + 1]) {
                break;
            }

            i++;
        }

        if (i == end) {
            return chars.slice(0, end + 1).join("");
        } else {
            for (let j = i; j <= end - 2; j++) {
                chars[j] = chars[j + 2];
            }

            end -= 2;
        }
    }

    return chars.slice(0, end + 1).join("");
}

// Attempt #2 - Custom Linked List (timeout for large strings)
class LinkedList {
    value: string;
    next: LinkedList | null;

    constructor(value: string, next: LinkedList) {
        this.value = value;
        this.next = next;
    }
}

function removeDuplicates2(s: string): string {
    let head = new LinkedList(s.charAt(0), null);
    let node = head;
    let prev = null;

    for (let i = 1; i < s.length; i++) {
        node.next = new LinkedList(s.charAt(i), null);
        node = node.next;
    }

    while (head?.next != null) {
        let modified = false;
        node = head;
        prev = null;

        while (node?.next != null) {
            let a = node.value;
            let b = node.next.value;

            if (a == b) {
                if (!prev) {
                    head = node.next.next;
                    node = head;
                } else {
                    prev.next = node.next.next;
                    node = prev.next;
                }

                modified = true;
            } else {
                prev = node;
                node = node.next;
            }
        }

        if (!modified) {
            break;
        }
    }

    let chars = [];

    while (head) {
        chars.push(head.value);
        head = head.next;
    }

    return chars.join("");
}

// Success - Stack (using an array)
function removeDuplicates3(s: string): string {
    let result = [];

    for (let c of s) {
        if (result[result.length - 1] == c) {
            result.pop();
        } else {
            result.push(c);
        }
    }

    return result.join("");
}